plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java

```diff
@@ -190,15 +224,26 @@ public ListenableFuture<Void> createEmptyFile(URI file)
     @Override
     public ListenableFuture<Void> deleteRecursively(URI dir)
     {
-        ImmutableList.Builder<String> keys = ImmutableList.builder();
-        return stats.getDeleteRecursively().record(transformFuture(Futures.transformAsync(
-                toListenableFuture((listObjectsRecursively(dir).subscribe(listObjectsV2Response ->
-                        listObjectsV2Response.contents().stream().map(S3Object::key).forEach(keys::add)))),
-                ignored -> {
-                    keys.add(keyFromUri(dir) + DIRECTORY_SUFFIX);
-                    return deleteObjects(getBucketName(dir), keys.build());
-                },
-                directExecutor())));
+        if (gcsClient.isPresent()) {
+            // GCS is not compatible with S3's multi-object delete API https://cloud.google.com/storage/docs/migrating#methods-comparison
+            return stats.getDeleteRecursively().record(asVoid(gcsDeleteExecutor.submit(() -> {
+                Storage storage = gcsClient.get();
+                StorageBatch batch = storage.batch();
+                Page<Blob> blobs = storage.list(getBucketName(dir), Storage.BlobListOption.prefix(keyFromUri(dir)));
+                for (Blob blob : blobs.iterateAll()) {
+                    batch.delete(blob.getBlobId());
```

Comment:
Are there any limits on batch size? Should we create multiple batches if `blobs` is to big?

Reply:
GCP Java JDK already handled for us https://github.com/googleapis/java-storage/blob/2fd3378eaffb34ed0bb1e6f377d81bb6ccecf030/google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java#L197-L210